README.md
# PHP Units of Measure
master: [![Build Status](https://travis-ci.org/landy2005/php-units-of-measure.png?branch=master)](https://travis-ci.org/landy2005/php-units-of-measure)

## Introduction
This is a PHP library for representing and converting physical units of measure.  The utility of this library is in encapsulating physical quantities in such a way that you don't have to keep track of which unit they're represented in.  For instance:

``` php
use PhpUnitsOfMeasure\PhysicalQuantity\Length;

$height = new Length(6.16, 'feet');
echo $height->toUnit('m');

// would print 1.87757, which is 6.16 feet in meters.
```

Having this abstraction allows you to create interfaces that accept physical quantities without requiring them to be in a particular unit.  For example, this function assumes the height is a float of a particular unit (presumably feet), and is therefore undesirably tied to a specific unit of measure:

``` php
// Tied to a specific unit of measure
function isTooTallToRideThisTrain( $height )
{
  return $height > 5;
}

// Calling the function requires that you first convert whatever quantity
// you have into the expected units:
isTooTallToRideThisTrain(2 / 0.3048);
```

Whereas this version using this library allows for height to be provided in whatever unit is convenient:

``` php
use PhpUnitsOfMeasure\PhysicalQuantity\Length;

// Free to operate on lengths in any unit of measure
function isTooTallToRideThisTrain( Length $height )
{
  return $height->toUnit('ft') > 5;
}

// Calling the function now allows any unit to be used:
isTooTallToRideThisTrain( new Length(2, 'm') );
```

## Installation
This library is best included in your projects via Composer.  See the [Composer website](https://getcomposer.org/) for more details, and see the [Packagist.org site for this library](https://packagist.org/packages/php-units-of-measure/php-units-of-measure).

If you'd prefer to manually include this library as a dependency in your project, then it is recommended that you use a [PSR-4](https://www.php-fig.org/psr/psr-4/) compliant PHP autoloader.  The mapping between this project's root namespace and its base directory is:
- vendor namespace 'PhpUnitsOfMeasure\' maps to the library's base directory 'source/'

See the documentation of your autoloader for further instructions.

### Project Tags and Versions
This project follows the guidelines set out in [Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html).  In general, versions are of the form 'X.Y.Z', and increments to X denote backward-incompatible major changes.

It is recommended that if your project includes this project as a dependency and you are using an automated dependency management tool such as [Composer](https://getcomposer.org/), then you should 'pin' the major version (X) and allow only variations in 'Y' (minor changes) and 'Z' (bugfixes).  See the documentation of your dependency manager for more details.


## Use
### Conversion
As in the examples above, the basic usage of this library is in representing physical quantities and converting between typical units of measure.  For example:

``` php
use PhpUnitsOfMeasure\PhysicalQuantity\Mass;

$quantity = new Mass(6, 'lbs');
echo $quantity->toUnit('g');
```
It's also possible to implicity cast a quantity to a string, which will display its original value:

``` php
use PhpUnitsOfMeasure\PhysicalQuantity\Mass;

$quantity = new Mass(6, 'pounds');
echo $quantity; // '6 lbs'
```

### Arithmetic Operators
There's also support for addition and subtraction.  The values of the physical quantity objects are immutable, and so these arithmetic methods return new quantity objects representing the results:

``` php
use PhpUnitsOfMeasure\PhysicalQuantity\Volume;

$first  = new Volume(6, 'liters');
$second = new Volume(6, 'cups');

$sum = $first->add($second);
echo $sum; // 7.4195292 l

$difference = $first->subtract($second);
echo $difference; // 4.5804708 l
```

### Adding new Units of Measure to Existing Quantities
Ocassionally, you will need to add a new unit of measure to a pre-existing quantity.

For example, let's say in a project you need a new measure of length, called "cubits".  You have two options: you can permanently add the new unit of measure to a new child class of the `\PhpUnitsOfMeasure\PhysicalQuantity\Length` class (or add it directly to that class and submit a pull request to get it added upstream, if appropriate), or you can add the unit temporarily at run time, inside your calling code.

#### Adding a New Unit of Measure at Runtime
To add a new unit of measure to an existing quantity at run time, you'd do this:

``` php
use PhpUnitsOfMeasure\PhysicalQuantity\Length;
use PhpUnitsOfMeasure\PhysicalQuantity\UnitOfMeasure;

// It's ok to create objects with cubits before the new unit is registered, since
// the conversion doesn't happen until an output method is called
$length = new Length(14, 'cubits');

// Build a new Unit of Measure object which represents the new unit, and which
// knows how to convert between the new unit and the quantity's native unit
// (in this case, meters).
$cubit = new UnitOfMeasure(

    // This is the official name of this unit - typically it's the standard
    // abbreviation
    'cb',

    // The second parameter is a function that converts from the native unit
    // to this unit
    function ($valueInNativeUnit) {
        return $valueInNativeUnit / 0.4572;
    },

    // The third parameter is a function that converts from this unit to the
    // native unit
    function ($valueInThisUnit) {
        return $valueInThisUnit * 0.4572;
    }
);

// Any alias names for this unit can be added here, to make it easier to use
// variations
$cubit->addAlias('cubit');
$cubit->addAlias('cubits');

// Register the new unit of measure with the quantity class
Length::addUnit($cubit);

// Now that the unit is registered, you can cast the measurement to any other
// measure of length
echo $length->toUnit('feet'); // '21'
```

##### Shorthand Factory Methods
Note that when creating instances of `UnitOfMeasure`, there are a couple of convenience static factory methods.  The first lets you instantiate units of measure which have linear scaling factors from the native unit. That is, the conversion function fits into the form `'Value in the native unit of measure' = 'Value in this unit of measure' * F`, where `F` is the scaling factor.

``` php
$megameter = UnitOfMeasure::linearUnitFactory('Mm', 1e6);
$megameter->addAlias('Megameter');
$megameter->addAlias('Megametre');
Length::addUnit($megameter);
```

The other convenience method is a special case of the above scaling factor factory method where the scaling factor is set to exactly 1, and serves as a convenient way of generating the native unit of measure.  All physical quantities must have one and only one native unit, so this method will probably only be called once per Physical Quantity class:

``` php
$meter = UnitOfMeasure::nativeUnitFactory('m');
$meter->addAlias('meter');
$meter->addAlias('metre');
Length::addUnit($meter);
```

##### Automatically Generating Metric Units
For units that use the metric system, there's a convenience trait available for classes which implement`PhysicalQuantityInterface` which will automatically generate the full continuum of metric units from a single unit.  For instance:

``` php
namespace PhpUnitsOfMeasure\PhysicalQuantity;

use PhpUnitsOfMeasure\AbstractPhysicalQuantity;
use PhpUnitsOfMeasure\UnitOfMeasure;
use PhpUnitsOfMeasure\HasSIUnitsTrait;

class Mass extends AbstractPhysicalQuantity
{
    use HasSIUnitsTrait;

    protected static $unitDefinitions;

    protected static function initialize()
    {
        // Kilogram
        $kilogram = UnitOfMeasure::nativeUnitFactory('kg');
        $kilogram->addAlias('kilogram');
        $kilogram->addAlias('kilograms');
        static::addUnit($kilogram);

        static::addMissingSIPrefixedUnits(
            $kilogram,
            1e-3,
            '%pg',
            [
                '%Pgram',
                '%Pgrams',
            ]
        );
    }
}
```

Here we're generating the native unit for mass, kilogram, adding it to the quantity as usual, and then using it to generate the spectrum of SI units by calling the `addMissingSIPrefixedUnits()` static method provided by the `HasSIUnitsTrait` trait.

Of note, the second parameter (1e-3) is denoting that while kilograms are the native unit for Mass, there's a factor of 1/1000 between the kilogram and the base metric unit of mass: the gram.  For units such as seconds or meters where the native unit for the physical quantity is also the base unit for the metric prefix system, this factor would be 1.

The 3rd and 4th parameters contain templates for the units' names and alternate aliases, respectively. The replacement strings '%p' and '%P' are used to denote the abbreviated and long-form metric prefixes.  For instance, '%pg' would generate the series `..., 'mg', 'cg', 'dg', 'g', ...`, while the template '%Pgram' would generate the series `..., 'milligram', 'centigram', 'decigram', 'gram', ...` .

#### Permanently Adding a New Unit of Measure to a Physical Quantity
The examples above for adding new units of measure to physical quantities allow you to register new units for the duration of the PHP execution, but are lost once execution terminates; it would be necessary to repeat this process every time you created a new program with `Length` measurements and wanted to use cubits.

A new unit of measure can be permanently added to a Physical Quantity class by essentially the same process as the one-time examples, only it would be done inside the initialize() method of the quantity class.  For example:

``` php
namespace PhpUnitsOfMeasure\PhysicalQuantity;

use PhpUnitsOfMeasure\AbstractPhysicalQuantity;
use PhpUnitsOfMeasure\UnitOfMeasure;

class Length extends AbstractPhysicalQuantity
{
    protected static $unitDefinitions;

    protected static function initialize()
    {
        // ...
        // ...
        // Here's all the pre-existing unit definitions for Length
        // ...
        // ...

        // Cubit
        $cubit = UnitOfMeasure::linearUnitFactory('cb', 0.4572);
        $cubit->addAlias('cubit');
        $cubit->addAlias('cubits');
        static::addUnit($cubit);
    }
}
```

Now any program which uses `Length` will start with the cubits unit already built in.  Note that here we used the more concise linear unit factory method, but the result is equivalent to the expanded form calling the `UnitOfMeasure` constructor, as used above.  Also, notice that the `static` keyword was used instead of the class name, though either would be acceptable in this case.

### Adding New Physical Quantities
[Physical quantities](https://en.wikipedia.org/wiki/Physical_quantity) are categories of measurable values, like mass, length, force, etc.

For physical quantities that are not already present in this library, it will be necessary to write a class to support a new one.  All physical quantities implement the `\PhpUnitsOfMeasure\PhysicalQuantityInterface` interface, typically extend the `\PhpUnitsOfMeasure\AbstractPhysicalQuantity` class, and typically have only an `initialize()` method which creates the quantity's units of measure.  See above for typical examples of physical quantity classes and of how to add new units to a quantity class.

Note that every physical quantity has a chosen "native unit" which is typically the SI standard.  The main point for this unit is that all of the quantity's other units of measure will convert to and from this chosen native unit.  It's important to be aware of a quantity's native unit when writing conversions for new units of measure.

### Adding new Aliases to Existing Units
It may come up that the desired unit of measure exists for a given physical quantity, but there's a missing alias for the unit.  For example, if you thought 'footses' was an obviously lacking alias for the `Length` unit 'ft', you could temporarily add the alias like this:

``` php
use PhpUnitsOfMeasure\PhysicalQuantity\Length;

// It's ok to use footses here, since the conversion doesn't happen
// until later
$length = new Length(4, 'footses');

// Fetch the unit of measure object that represents the 'ft' unit
$footUnit = Length::getUnit('ft');

// Any alias names for this unit can be added here, to make it easier
// to use variations
$footUnit->addAlias('footses');

// Now that the unit has been modified with its new alias, you can cast
// the measurement to any other measure of length
echo $length->toUnit('m'); // '1.2192'
```

And of course, if you need to add the alias permanently, you can do so in the initialize() method of the quantity's class, as shown above.

## Testing and Contributing
Pull requests are welcome, especially regarding new units of measure or new physical quantities.  However, please note that there are many sources for conversion factors, and not all are careful to respect known precision.

In the United States, the standards body for measurement is NIST, and they've published [NIST Special Publication 1038 "The International System of Units (SI) - Conversion factors for General Use"](http://www.nist.gov/pml/wmd/metric/upload/SP1038.pdf).  This guide contains the approved conversion factors between various units and the base SI units.

Also note that any new physical quantities should have the appropriate SI unit chosen for their native unit of measure.

### Pull Requests and Merging
The workflow for this repository goes as follows:
- To develop new contributions, fork or branch from the `master` branch of the main repository
- Pull requests and contribution merges are always made to the `master` branch of the main repository
- From time to time, commits of `master` are tagged and a new version is released
- At present, there is no support for maintaining bug-fix branches of older project versions.  This is something we can revisit if a need arises.

End users of this repository should only use tagged commits in production.  Users interested in the current 'soon-to-be-released' code may use `master`, with the understanding that it may change unexpectedly.  All other existing branches (if any) should be considered 'feature' branches in development, and not yet ready for use.

### Local Testing Environment
There's a Vagrant virtual machine configuration included which is suitable for running the necessary unit tests.  To bring up the machine, make sure you have Vagrant and Virtualbox installed, and from the project root directory:


``` bash
vagrant up
vagrant ssh
cd /project
```

### Setting Up for Testing
The virtual machine development environment already has Composer installed.  Once you're ssh'ed into the virtual machine, install this project's dev dependencies:

``` bash
rm -rf vendor
composer.phar update --verbose --prefer-dist
```

### Unit Tests
All the tests associated with this project can be manually run with:

``` bash
composer test
```

### CodeSniffer
Codesniffer verifies that coding standards are being met.  Once the project is built with development dependencies, you can run the checks with:

``` bash
composer phpcs
```

### Continuous Integration
The above tests are automatically run against Github commits with Travis-CI.
- https://travis-ci.org/landy2005/php-units-of-measure


CHANGELOG.md
## v2.2.2 (November 3nd, 2023)
- Added ronna (R) and quetta (Q) si prefixes.
- Added IEC prefixed units support for Storage
- Added Storage units

## v2.2.1 (November 2nd, 2023)
- Drop support of php < 7.2

## v2.2.0 (February 17th, 2021)
- Added support for php 8.0
- Added option power factor for Si factor, ie for square (2) and cubic (3)
- Added new Physical Quantity volume flow with commonly used units
- Added lowercase aliases for temperature units
- Added Si prefixes for area unit square meter (m^2)
- Added Si prefixes for volume units cubic meter (m^3) and litres
- Added Si prefixes for velocity units m/s, m/min and m/h
- Added Si prefixes for acceleration unit m/s^2
- Additional units for energy: electronvolt, calorie
- Additional units for length: light year, parsec
- Additional units for mass: Hundredweight, US short Ton, US long Ton, Asian picul
- Additional units for pressure: Torr, Technical atmosphere, Kilo and Mega pound per Square Inch
- Additional units for velocity: Mach, meters/min, feet per minute, miles per second
- Additional unit for power: dBm
- Fixed native unit for mass as gram

## v2.1.0 (July 24th, 2016)
- Added getUnitDefinitions() method to PhysicalQuantity classes, to get a raw list of UnitofMeasure objects defined on that quantity
- Added UPPERCASE templating support for autogenerated metric units to support names like MEGA_METERS, for example.
- New Quantity physical quantity, with `mole` as the standard unit, to represent the Amount of Substance SI base quantity.
- New SolidAngle physical quantity, with `steradian` as the standard unit, to represent the solid angle quantity
- Additional aliases for Area units
- Additional units for time: decade, century, and millenium
- Additional units for volume: teaspoon, tablespoon, gallon, quart, fluid ounce, and pint

## v2.0.1 (December 12th, 2015)
- Migrated package from triplepoint/php-units-of-measure to php-units-of-measure/php-units-of-measure.
- New physical quantity: Energy
- New units of measure:
  - Area: are and decare
  - Length: nautical mile, mil, AU
  - Mass: stone, also slightly changed definitions of lb and oz
  - Time: yr, and jyr
  - Velocity: km/h, ft/s, mph, knot
- Automated CI testing now includes HHVM and PHP 7
- Removed the heavy vagrant git submodule, it was a silly thing.
- Expanded the Readme for better contribution and installation instructions.

## v2.0.0 (March 14th, 2015)
- Registered units of measure are now properties of PhysicalQuantity classes, and not individual instances of those classes.  As such, registering a new unit with a given PhysicalQuantity will make that unit immediately available to all inntances of that class.
   - PhysicalQuantity classes no longer define their units in their constructor, and instead have a new initialize() static method
   - The previously-existing registerUnitOfMeasure() method on physical quantity objects has been replaced with a new static addUnit() method
   - The HasSoUnitsTrait method addMissingSIPrefixedUnits is now static
- Added the toNativeUnit() method, to return values in their native unit of measure
- Added the isEquivalentQuantity() method, to support future situations where it's not obvious whether two physical quantities represent the same quantity type
- getSupportedUnits() is no longer available on physical quantities
- The PhysicalQuantity parent class is now named AbstractPhysicalQuantity
- Add a new DemonstrationTests test file, to demonstrate and test typical use cases
- All library exceptions extend from AbstractPhysicalQuantityException, making catching easier
- Added an interface for PhysicalQuantities, to support future work where not all physical quantity classes necessarily have the same parent

## v1.3.3 (March 15th, 2015)
- Loosened Composer requirement on PHP version to >=5.3.0 (was previously 5.3.5).

## v1.3.1 (August 23rd, 2014)
- Added information in the README about the new SI prefix units generation

## v1.3.0 (August 23rd, 2014)
- Add support for automatically generating metric-prefixed units from a single base unit
- Throw an exception on registering units of measure with names or aliases that collide with existing units on that quantity
- Disallow values that aren't numerical and unit names that aren't strings (ie, type checking)
- Add new method getSupportedUnits() to Physical Quantities
- Add new method getAliases() to Units of Measure
- Add arcminute and arcsecond as angle units
- Add all SI prefix units to meters, kilograms, seconds, amps, kelvin, candela, pascal, bar, radians, degrees, and arcseconds

## v1.2.0 (August 16th, 2014)
- Drop support for PHP 5.3
- Moved to PSR-4 namespace convention
- Drop support for an array of names passed in the constructor as a way to specify aliases for UnitOfMeasure objects; addAlias() now required.
- Add UnitOfMeasure::nativeUnitFactory() and UnitOfMeasure::linearUnitFactory() convenience factory methods
- Add optional spelling for 'metres' and 'litres'
- Add Vagrant-based testing virtual machine, for convenience

## v1.1.2 (February 9th, 2014)
 - New units of measure on Area: hectares, international acres
 - New units of measure on Temperature: Rankin, Delisle, Newton, Réaumur, Rømer
 - Improved unit test coverage

## v1.1.1 (July 8th, 2013)
 - New physical quantity: Angle
 - Additional units of measure on Time

## v1.1.0 (May 7th, 2013)
 - Add add() and subtract() methods to physical quantities
 - toString() on quantities now shows the original unit, as expected

## v1.0.0 (May 7th, 2013)
Initial Commit


source/HasSIUnitsTrait.php
<?php
namespace PhpUnitsOfMeasure;

/**
 * Physical quantities with this trait
 * have units which are metric and therefore have
 * a standard set of prefixes.
 */
trait HasSIUnitsTrait
{
    /**
     * Given the patterns for generating the names and aliases,
     * generate the various metric units of measure and add
     * them to this physical quantity.
     *
     * Names and Aliases are created by replacing identifiers in the respective
     * patterns.  The current allowed replacement identifiers are:
     *   %p = the abbreviated SI prefix, like 'M' for 'megameter' or 'k' for 'kilogram'
     *   %P = the full SI prefix, like 'mega' for 'megameter' or 'kilo' for 'kilogram'
     *   %U = uppercase version of %P
     *
     * So for instance, in order to generate 'kg', 'mg', and 'g' names for SI units, the
     * appropriate pattern would be '%pg'.  Similarly, to generate 'kilogram', 'milligram',
     * and 'gram' aliases, the pattern would be '%Pgram'.
     *
     * The $siUnit given in the 1st parameter must be some SI unit in the series of units
     * to be generated by this method.  This value is necessary to establish a conversion
     * factor between this continuum of SI units and the Physical Quantity's native unit.
     *
     * The second parameter provides a scaling factor between the given SI unit in the first parameter
     * and the base unit of the SI continuum (ie, 'grams', 'meters', 'seconds', etc).  For instance,
     * if a Kilogram unit of measure was passed for the 1st parameter, it would be necessary to
     * then pass 1e-3 in the 2nd parameter to indicate that a gram is 1/1000 of the given unit.
     *
     * @param  UnitOfMeasure $siUnit             A unit in this physical quantity that is an SI unit of measure
     * @param float          $toBaseSiUnitFactor The power-of-ten factor that converts the given SI unit into the not-prefixed SI base unit (ie 1e-3 for kilograms)
     * @param string         $namePattern        The pattern to apply to the base unit's name to generate a new SI unit name
     * @param  array         $aliasPatterns      The collection of alias patterns to use in generating a new SI unit's aliases
     * @param integer|null   $powerFactor        Use power factor for squares, qubic and other multiplication of SI factor (ie. square is 2, qubic is 3)
     */
    protected static function addMissingSIPrefixedUnits(
        UnitOfMeasure $siUnit,
        float         $toBaseSiUnitFactor,
        string        $namePattern,
        array         $aliasPatterns = [],
        int           $powerFactor = null
    ) {
        /**
         * The standard set of SI prefixes
         */
        $siPrefixes = [
            [
                'abbr_prefix' => 'Q',
                'long_prefix' => 'quetta',
                'factor'      => 1e30
            ],
            [
                'abbr_prefix' => 'R',
                'long_prefix' => 'ronna',
                'factor'      => 1e27
            ],
            [
                'abbr_prefix' => 'Y',
                'long_prefix' => 'yotta',
                'factor'      => 1e24
            ],
            [
                'abbr_prefix' => 'Z',
                'long_prefix' => 'zetta',
                'factor'      => 1e21
            ],
            [
                'abbr_prefix' => 'E',
                'long_prefix' => 'exa',
                'factor'      => 1e18
            ],
            [
                'abbr_prefix' => 'P',
                'long_prefix' => 'peta',
                'factor'      => 1e15
            ],
            [
                'abbr_prefix' => 'T',
                'long_prefix' => 'tera',
                'factor'      => 1e12
            ],
            [
                'abbr_prefix' => 'G',
                'long_prefix' => 'giga',
                'factor'      => 1e9
            ],
            [
                'abbr_prefix' => 'M',
                'long_prefix' => 'mega',
                'factor'      => 1e6
            ],
            [
                'abbr_prefix' => 'k',
                'long_prefix' => 'kilo',
                'factor'      => 1e3
            ],
            [
                'abbr_prefix' => 'h',
                'long_prefix' => 'hecto',
                'factor'      => 1e2
            ],
            [
                'abbr_prefix' => 'da',
                'long_prefix' => 'deca',
                'factor'      => 1e1
            ],
            [
                'abbr_prefix' => '',
                'long_prefix' => '',
                'factor'      => 1
            ],
            [
                'abbr_prefix' => 'd',
                'long_prefix' => 'deci',
                'factor'      => 1e-1
            ],
            [
                'abbr_prefix' => 'c',
                'long_prefix' => 'centi',
                'factor'      => 1e-2
            ],
            [
                'abbr_prefix' => 'm',
                'long_prefix' => 'milli',
                'factor'      => 1e-3
            ],
            [
                'abbr_prefix' => 'µ',
                'long_prefix' => 'micro',
                'factor'      => 1e-6
            ],
            [
                'abbr_prefix' => 'n',
                'long_prefix' => 'nano',
                'factor'      => 1e-9
            ],
            [
                'abbr_prefix' => 'p',
                'long_prefix' => 'pico',
                'factor'      => 1e-12
            ],
            [
                'abbr_prefix' => 'f',
                'long_prefix' => 'femto',
                'factor'      => 1e-15
            ],
            [
                'abbr_prefix' => 'a',
                'long_prefix' => 'atto',
                'factor'      => 1e-18
            ],
            [
                'abbr_prefix' => 'z',
                'long_prefix' => 'zepto',
                'factor'      => 1e-21
            ],
            [
                'abbr_prefix' => 'y',
                'long_prefix' => 'yocto',
                'factor'      => 1e-24
            ],
        ];

        // Determine the conversion factor from the no-prefix SI unit to the physical quantity's native unit
        $noPrefixToNativeUnitFactor = $siUnit->convertValueToNativeUnitOfMeasure(1) * $toBaseSiUnitFactor;

        // For each of the standard SI prefixes, attempt to register a new unit of measure
        foreach ($siPrefixes as $prefixDefinition) {
            // Build a function for resolving a pattern into a unit name
            $parsePattern = function ($pattern) use ($prefixDefinition) {
                return strtr(
                    $pattern,
                    [
                        '%p' => $prefixDefinition['abbr_prefix'],
                        '%P' => $prefixDefinition['long_prefix'],
                        '%U' => strtoupper($prefixDefinition['long_prefix'])
                    ]
                );
            };

            // Generate the base name of the new unit
            $name = $parsePattern($namePattern);

            // Determine the factor that converts the new unit into the physical quantity's
            //   native unit of measure.
            if (is_int($powerFactor) && $powerFactor !== 0) {
                $prefixDefinition['factor'] = $prefixDefinition['factor'] ** $powerFactor;
            }
            $toNativeUnitFactor = $noPrefixToNativeUnitFactor * $prefixDefinition['factor'];

            // Instantiate the new unit of measure
            $newUnit = UnitOfMeasure::linearUnitFactory($name, $toNativeUnitFactor);

            // Generate the aliases of the new unit
            foreach ($aliasPatterns as $aliasPattern) {
                $newUnitAlias = $parsePattern($aliasPattern);
                $newUnit->addAlias($newUnitAlias);
            }

            // If the unit doesn't conflict with any of the already-existing units, register it
            if (!static::unitNameOrAliasesAlreadyRegistered($newUnit)) {
                static::addUnit($newUnit);
            }
        }
    }
}
